---
title: "CentralDogmaFun"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{CentralDogmaFun}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(CentralDogmaFun)
```

### Sofie

### DNA_to_RNA

The DNA_to_RNA function convert a DNA sequence into an RNA sequence by replacing thymine (T) with uracil (U) in the DNA input sequence. The process mimics the biological process of transcription, where DNA is transcribed to RNA, and thymine is substituted with uracil \\ In the DNA_to_RNA function a DNA sequence is taken as input, this is represented by a string of nucleotide letters (A, T, C and G). using the gsub function, all occurences of "T" in the DNA sequence are replaced with "U" to reflect the substitution in nucleotides that happens during transcription. The resulting RNA sequence is then returned as output. \\ If "ATTGCG" is set as input, the function will return "AUUGCG". See the code for reference:

```{r}
DNA_to_RNA("ATTGCG")
```



### RNA_to_codon

The RNA_to_codon function converts a string of RNA seuqence into codons of 3 nucleotides. The procces mimics how the nucelotides of the mRNA sequence are being read into codons which then eventually is translated into amino acids in the portein seuqence. \\ The function takes as input a RNA sequence in a string plus the parameter start, which should be set to a default of = 1, unless another starting point than the beginning of the RNA sequence is wanted. The function then seperates the RNA string into a vector of strings containing the codon in each string. \\ If "UGGUCC" is set as input, e.i. RNA_to_codon("UGGUCC", start=1) the function will return c("UGG","UCC") as output where "UGG" and "UCC" is each a codon as the function begins at position 1. 

```{r}
RNA_to_codon("UGGUCC", start=1) # returns c("UGG","UCC")
```

### codons_to_aa

The codons_to_aa function converts codons into amino acids. The process mimics the biological process of translation, where RNA is read in codons by the ribosome and the corrsponding amino acid is added to the growing peptide chain. \\ The function takes a vector of codons as strings as input. The strings are converted into the corresponding one-letter amino acid code using the codon table. The one-letter amino acid codes are combined into one string and returned as output. \\ If c("UUU","CAA","GAC") is set as input, e.i. codons_to_aa(c("UUU","CAA","GAC")) the function will return "FQD" as output.

```{r}
codons_to_aa(c("UUU","CAA","GAC")) # returns "FDQ"
``` 

### protein_plot

The protein_plot function takes a sequence of aminoacids (AA), counts the amount of each AA and plot them in a column-diagram/histogram. This is done in order to give an overview of the content of an AA-seq. The function works by first finding all unique AA in the sequence.
Hereafter all the different proteins are counted in categories of the unique proteins. These counts are then used in a ggplot2::geom_col plot. Hereby the function is able to costumize the amount of columns (dependent on the unique AA's). Hereby, this function is able to give a visualization of the contents in the end product of the central dogma.

```{r}
protein_plot("FDQQDQKA") #plots a histogram of the contents. For this sequence there will be 5 columns: F,D,Q,K,A with the counts: 1,2,3,1,1.
```



### Use cases for the package
The package is intended to easily transcribe and translate a random or known DNA sequence, resembling the steps in the central dogma. The functions in the package are made so that they can be used sequentially in a pipeline and fully translate the DNA into protein and plot the count of the different amino acids. 

### Discussion on limiting the number of dependencies
When making code more useable, it is crucial to limit the number of dependencies. \\ The more dependencies a code is relying on, the more points of potential failture of your code arise. This is a consequence of code/packages being uploaded and then not updated. If a code of yours is then built, and therefore dependent, on this package, it might potneitally stop wokring, if said package is not updated along other packages used in yoru code which might depend on each other. 

### Differences between loading packages
One can call functions and packages in different manners. In general one can use a package by either loading or attaching. When loading the package, you connect the specific function to the library you want to use it from. If one instead uses attaching, you simply load the library and then call then function when needed in your code. \\ One might say, that attahcing makes it easier to call functions, however it also adds to riks of conflict if a function names is present in more than one of the libraries loaded. 
